repair/repair.cc

```diff
@@ -71,17 +71,17 @@ node_ops_metrics::node_ops_metrics(shared_ptr<repair::task_manager_module> modul
     auto ops_label_type = sm::label("ops");
     _metrics.add_group("node_ops", {
         sm::make_gauge("finished_percentage", [this] { return bootstrap_finished_percentage(); },
-                sm::description("Finished percentage of node operation on this shard"), {ops_label_type("bootstrap")}),
+                scylla_node_ops_finished_percentage , {ops_label_type("bootstrap")}),
         sm::make_gauge("finished_percentage", [this] { return replace_finished_percentage(); },
-                sm::description("Finished percentage of node operation on this shard"), {ops_label_type("replace")}),
+                scylla_node_ops_finished_percentage, {ops_label_type("replace")}),
         sm::make_gauge("finished_percentage", [this] { return rebuild_finished_percentage(); },
```

Comment:
This generates more duplication - the name `scylla_node_ops_finished_percentage` has one definition and one use.

Can't we add markup so that the source file can be processed directly?

Reply:
good idea, probably we can embed the yaml above as comment with a dedicated mark at the header of the file, so our tool can skip the ones without the mark.

Reply:
Actually in his example this variable has one definition but several uses.
And the separate definition and use has the extra benefit that if there is a definition (of a **static** variable) that isn't used, the compiler warns about it. This means that a tool that reads just the ".hh" file to generate documentation (this tool wouldn't need to read the .cc file) knows for sure that the lines it sees are really used in the code.

What bothered me is that nothing checks the variable name "scylla_node_ops_finished_percentage" fits the name of the metric, so a tool that only reads the .hh file can't be sure the metrics names are correct. I think this should be solved by a macro, that verifies that the new gauge's full name is equal to the string (`##` C preprocessor trick) of the variable's name.

Reply:
In the general case it's one use per definition.

And it there is no extra file, the tool that reads the source is also sure that the lines are really used in the code.

Reply:
I'm sure we can move the entire metric_group definition (except the lambda) to static context, then the dynamic context just adds the lambda.

Then we can have a `scylla describe-metrics` command that dumps all the static content.

Reply:
> Can't we add markup so that the source file can be processed directly?

> good idea, probably we can embed the yaml

Markup? Yaml? What are we, web developers?

Behold:

```
#define DEFINE_METRIC(NAME_RAW, DESC_RAW, EXPR) \
[&] () -> decltype(auto) { \
    __attribute__((section(".metrics"))) \
    __attribute__((used)) \
    constexpr static const char NAME_RAW[] = DESC_RAW; \
    constexpr static const char* NAME = #NAME_RAW; \
    constexpr static const char* DESC = NAME_RAW; \
    return EXPR; \
}()
[...]
DEFINE_METRIC(node_ops_finished_percentage, "Finished percentage of node operation on this shard", sm::make_gauge(NAME, [this] { return bootstrap_finished_percentage(); }), sm::description(DESC), {ops_label_type("bootstrap")});
```

And now, to list all metrics, you can just iterate over Scylla's `.symtab` and write out all names/descriptions with `st_shndx` equal to `.metrics`.

Reply:
This is similar to my second idea, though I wouldn't stoop so low as to use the preprocessor.

Using a section instead of executing the binary is a good improvement.

Reply:
> though I wouldn't stoop so low as to use the preprocessor

@michoecho have you every heard a song by Gotye: "Somebody That I Used to Know"?

:musical_note: **No, you didn't have to stoop so low** :musical_note:
:musical_note: Have your _preprocessor_ collect your records and then change your number :musical_note:
:musical_note: I guess that I don't need that, though :musical_note:
:musical_note: Now you're just somebody that I used to know... :musical_note:

Reply:
The duplication is on purpose. Having the full metric name in the code makes it easier to search for a metric in the code (as a developer).

Having a file with only metrics names and descriptions (and yes, I thought about using Marcro to remove the boilerplate but never thought anyone would go along) makes it easier for documentation review doc generation.

Reply:
It's just making everything more complicated and error prone. Its like with C++ declarations and definitions, everyone ends up doing everything inline because it's such a pain to duplicate things.

Let's try to find a solution that doesn't cause duplication.

Reply:
We are starting from an inline situation. What I'm trying to achieve:
1. make sure we have 1-1 mapping between the description and the actual code, so if someone adds or removes a metric, the list of metrics description will be up-to-date
2. do not rely on runtime, this is what we have today, and it does not cover all metrics
3. make it easy for IDE

My assumption is that after the first phase, that I would do (90% using Python scripts and a few abominations I'll deal with by hand), the number of new metrics per release will be relatively low and would be easy to handle.

Reply:
> We are starting from an inline situation. What I'm trying to achieve:
> 
> 1. make sure we have 1-1 mapping between the description and the actual code, so if someone adds or removes a metric, the list of metrics description will be up-to-date

That already works now.

> 2. do not rely on runtime, this is what we have today, and it does not cover all metrics

That's fixed by moving the definitions to static scope though it necessarily creates some duplication).

> 3. make it easy for IDE

What about us poor programmers?


> 
> My assumption is that after the first phase, that I would do (90% using Python scripts and a few abominations I'll deal with by hand), the number of new metrics per release will be relatively low and would be easy to handle.

Reply:
> > We are starting from an inline situation. What I'm trying to achieve:
> > 
> > 1. make sure we have 1-1 mapping between the description and the actual code, so if someone adds or removes a metric, the list of metrics description will be up-to-date
> 
> That already works now.

But we don't have a way of knowing what metrics we have, you cannot grep a metric name, and the documentation has no way of getting a list of metric->description per version.

> 
> > 2. do not rely on runtime, this is what we have today, and it does not cover all metrics
> 
> That's fixed by moving the definitions to static scope, though it necessarily creates some duplication).

We can unify the header file with a static part in the source file (we all know it's basically the same in C++) the downside is the doc team, if they will have their own files (not many, I think we have around 50 files with metrics in them) they can own them and it will make it easier for them to send PR or set as reviewers.

> 
> > 3. make it easy for IDE
> 
> What about us poor programmers?

You'll get code completion, and compile time errors, I'll do the heavy lifting of moving all the current code to the new format, and think about the happy maintainers who could assign someone from the doc team whenever someone changes a description file.

> 
> > My assumption is that after the first phase, that I would do (90% using Python scripts and a few abominations I'll deal with by hand), the number of new metrics per release will be relatively low and would be easy to handle.